storage/scan_queue.go

```diff
@@ -0,0 +1,145 @@
+	// write intents. If this much time has passed since the last scan
+	// and write intents are present, the range should be queue. Cleaning
+	// up intents allows transaction records to be GC'd.
+	intentSweepInterval = 10 * 24 * time.Hour // 10 days
+	// intentAgeThreshold is the threshold after which an extant intent
+	// will be resolved.
+	intentAgeThreshold = 1 * time.Hour // 1 hour
+	// verificationInterval is the target duration for verifying on-disk
+	// checksums via full scan.
+	verificationInterval = 30 * 24 * time.Hour // 30 days
+)
+
+// scanQueue manages a queue of ranges slated to be scanned in their
+// entirety using the MVCC versions iterator. Currently, range scans
+// manage the following tasks:
+//
+//  - GC of version data via TTL expiration (and more complex schemes
+//    as implemented going forward).
+//  - Resolve extant write intents and determine oldest non-resolvable
+//    intent.
+//  - Periodic verification of on-disk checksums to identify bit-rot
+//    in read-only data sets. See http://en.wikipedia.org/wiki/Data_degradation.
+//
+// The shouldQueue function combines the need for all three tasks into
+// a single priority. If any task is overdue, shouldQueue returns true.
+type scanQueue struct {
+	*baseQueue
+}
+
+// newScanQueue returns a new instance of scanQueue.
+func newScanQueue() *scanQueue {
+	sq := &scanQueue{}
+	sq.baseQueue = newBaseQueue(sq.shouldQueue, sq.process, scanQueueMaxSize)
+	return sq
+}
+
+// shouldQueue determines whether a range should be queued for
+// scanning, and if so, at what priority. Returns true for shouldQ in
+// the event that there are GC'able bytes, or it's been longer since
+// the last scan than the intent sweep or verification
+// intervals. Priority is derived from the addition of priority from
+// GC'able bytes and how many multiples of intent or verification
+// intervals have elapsed since the last scan.
+func (sq *scanQueue) shouldQueue(now time.Time, rng *Range) (shouldQ bool, priority float64) {
+	scanMeta, err := rng.GetScanMetadata()
+	if err != nil {
+		log.Errorf("unable to fetch scan metadata: %s", err)
+		return
+	}
+	elapsedNanos := now.UnixNano() - scanMeta.LastScanNanos
+
+	// Compute non-live bytes.
+	bytes, err := engine.GetRangeSize(rng.rm.Engine(), rng.Desc.RaftID)
+	if err != nil {
+		log.Errorf("unable to fetch range size stats: %s", err)
+		return
+	}
+	liveBytes, err := engine.GetRangeStat(rng.rm.Engine(), rng.Desc.RaftID, engine.StatLiveBytes)
+	if err != nil {
+		return
+	}
+	nonLiveBytes := bytes - liveBytes
+
+	// GC score.
+	estGCBytes := scanMeta.GC.EstimatedBytes(elapsedNanos, nonLiveBytes)
+	gcScore := float64(estGCBytes) / float64(gcByteCountNormalization)
+
+	// Intent sweep score. First check for intents. We only compute an
+	// intent score if there are any outstanding intents.
+	intentBytes, err := engine.GetRangeStat(rng.rm.Engine(), rng.Desc.RaftID, engine.StatIntentBytes)
+	if err != nil {
+		return
```

Comment:
I'm adding error logging for these cases and ignoring the contribution to the score. Good catch.